pkg/storage/chunk/client/ibmcloud/cos_object_client_test.go

```diff
@@ -518,3 +529,74 @@ func Test_APIKeyAuth(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, resp, strings.Trim(string(data), "\n"))
 }
+
+func Test_TrustedProfileAuth(t *testing.T) {
+	testToken := "test"
+	tokenType := "Bearer"
+	resp := "testGet"
+
+	cosSvr := cosServer(testToken, tokenType, resp)
+	defer cosSvr.Close()
+
+	authServer := authServer(testToken, tokenType)
+	defer authServer.Close()
+
+	file, err := createTempFile("crtoken", "test cr token")
+	require.NoError(t, err)
+	defer os.Remove(file.Name())
+
+	cosConfig := COSConfig{
+		BucketNames:        "dummy",
+		Endpoint:           cosSvr.URL,
+		Region:             "dummy",
+		ForcePathStyle:     true,
+		AuthEndpoint:       authServer.URL,
+		CRTokenFilePath:    file.Name(),
+		TrustedProfileName: "test",
+		BackoffConfig: backoff.Config{
+			MaxRetries: 1,
+		},
+	}
+
+	cosClient, err := NewCOSObjectClient(cosConfig, hedging.Config{})
+	require.NoError(t, err)
+
+	reader, _, err := cosClient.GetObject(context.Background(), "key-1")
+	require.NoError(t, err)
+
+	data, err := io.ReadAll(reader)
+	require.NoError(t, err)
+	require.Equal(t, resp, strings.Trim(string(data), "\n"))
+}
+
+func cosServer(accessToken, tokenType, resp string) *httptest.Server {
+	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Header.Get("Authorization") != fmt.Sprintf("%s %s", tokenType, accessToken) {
+			w.WriteHeader(http.StatusUnauthorized)
+			return
+		}
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintln(w, resp)
+	}))
+}
+
+func authServer(accessToken, tokenType string) *httptest.Server {
+	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		token := token.Token{
+			AccessToken:  accessToken,
+			RefreshToken: "test",
+			TokenType:    tokenType,
+			ExpiresIn:    int64((time.Hour * 24).Seconds()),
+			Expiration:   time.Now().Add(time.Hour * 24).Unix(),
+		}
+
+		data, err := json.Marshal(token)
+		if err != nil {
+			w.WriteHeader(http.StatusUnauthorized)
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+		w.Write(data)
```

Comment:
Better to handle the `Write` error and say `StatusUnauthorized` if any error.

Reply:
It's something we can't do, the problem is if we call `WriteHeader` after the `Write` they will show a warning. As far as I understood it's okay to skip error check in this case. What maximum we can do is log the error, since it's test casae I don't think we have to do that.

Reply:
My doubt is that we might get lint error as we are not capturing the error and handling it. Try running lint once from Loki Makefile.

Reply:
Agree with @adityacs , that will make the linting fail.

Reply:
Linter was not complaining for me when I ran it locally. Also I have seen that people adding exception for function like `w.Write(data)` and  `fmt.Fprintln(w, resp)` since sometimes checking error for them deos not make any sense.

As I explained above if we call `WriteHeader` after the `Write`, we will get a warining when we run the test case which says that `WriteHeader` statement is unnecessary. Eg.., below:

```
2023/03/23 13:42:50 http: superfluous response.WriteHeader call from github.com/grafana/loki/pkg/storage/chunk/client/ibmcloud.authServer.func1 (cos_object_client_test.go:600
```

Reply:
if you open the PR to Loki, it would complain, the same issue happened with the other PR.

Reply:
We are checking the error and logging it now.